modules/core/src/main/java/org/locationtech/jts/algorithm/hull/ConcaveHull.java

```diff
@@ -0,0 +1,718 @@
+    hull.setMaximumAreaRatio(areaRatio);
+    return hull.getHull();
+  }
+  
+  private Geometry inputGeometry;
+  private double maxEdgeLength = 0.0;
+  private double maxEdgeLengthFactor = -1;
+  private double maxAreaRatio = 0.0; 
+  private boolean isHolesAllowed = false;
+  private GeometryFactory geomFactory;
+
+
+  /**
+   * Creates a new instance for a given geometry.
+   * 
+   * @param geom the input geometry
+   */
+  public ConcaveHull(Geometry geom) {
+    this.inputGeometry = geom;
+    this.geomFactory = geom.getFactory();
+  }
+  
+  /**
+   * Sets the target maximum edge length for the concave hull.
+   * A value of 0.0 produces a concave hull of minimum area
+   * that is still connected.
+   * The {@link #uniformGridEdgeLength(Geometry)} may be used as
+   * the basis for estimating an appropriate target maximum edge length.
+   * 
+   * @param edgeLength a non-negative length
+   * 
+   * @see #uniformGridEdgeLength(Geometry)
+   */
+  public void setMaximumEdgeLength(double edgeLength) {
+    if (edgeLength < 0)
+      throw new IllegalArgumentException("Edge length must be non-negative");
+    this.maxEdgeLength = edgeLength;
+    maxEdgeLengthFactor = -1;
+  }
+  
+  /**
+   * Sets the target maximum edge length factor for the concave hull.
+   * The edge length factor is a fraction of the length delta
+   * between the longest and shortest edges 
+   * in the Delaunay Triangulation of the input points. 
+   * A value of 1.0 produces the convex hull. 
+   * A value of 0.0 produces a concave hull of minimum area
+   * that is still connected.
+   * 
+   * @param edgeLengthFactor a length factor value between 0 and 1
+   */
+  public void setMaximumEdgeLengthFactor(double edgeLengthFactor) {
+    if (edgeLengthFactor < 0 || edgeLengthFactor > 1)
+      throw new IllegalArgumentException("Edge length ratio must be in range [0,1]e");
+    this.maxEdgeLengthFactor = edgeLengthFactor;
+  }
+  
+  /**
+   * Sets the target maximum concave hull area as a ratio of the convex hull area.
+   * A value of 1.0 produces the convex hull 
+   * (unless a maximum edge length is also specified).
+   * A value of 0.0 produces a concave hull with the smallest area
+   * that is still connected.
+   * 
+   * @param areaRatio a ratio value between 0 and 1
+   */
+  public void setMaximumAreaRatio(double areaRatio) {
+    if (areaRatio < 0 || areaRatio > 1)
+      throw new IllegalArgumentException("Area ratio must be in range [0,1]");
+    this.maxAreaRatio = areaRatio;
+  }
+  
+  /**
+   * Sets whether holes are allowed in the concave hull polygon.
+   * 
+   * @param isHolesAllowed true if holes are allowed in the result
+   */
+  public void setHolesAllowed(boolean isHolesAllowed) {
+    this.isHolesAllowed = isHolesAllowed;
+  }
+  
+  /**
+   * Gets the computed concave hull.
+   * 
+   * @return the concave hull
+   */
+  public Geometry getHull() {
+    if (inputGeometry.isEmpty()) {
+      return geomFactory.createPolygon();
```

Comment:
If the input geometry is a POLYGON Z, the output should respect the dimensionality. (Yeah.)

Reply:
Yes, I realize this is something that needs work.  Will do that soon.